### DISCUSSION TITLE
(non) Sunton 7" TFT 800*480 Display (ESP32-8048S070)

### DISCUSSION THREAD

**Discussion by ChrisVonFrosta** [arendst/Tasmota#20560] [Q&A]
While I'm waiting/hoping for support for another display in this [thread](https://github.com/arendst/Tasmota/discussions/20558), I've been working with another display. This one [here](https://www.aliexpress.com/item/1005005099968475.html?spm=a2g0o.order_list.order_list_main.5.61221802ml4Fho).

 I set this to Sunton 7 inch via auto-configuration and also wanted to start here with the command "UrlFetch https://raw.githubusercontent.com/arendst/Tasmota/development/tasmota/berry/haspmota/haspmota_demo.tapp". But strangely, the command "UrlFetch" is unknown. `
At least the download with "UrlFetch" works on the board in the thread linked above. On the other hand, the BerryScript of the examples (here and here) can be executed without errors, which returns an error in the thread linked above. I don't understand that. Unfortunately, nothing appears on the display. How do I proceed with this display?

```
00:00:00.002 HDW: ESP32-S3 v0.1 
00:00:00.033 UFS: FlashFS mounted with 12588 kB free
00:00:00.043 CFG: Loaded from File, Count 22
00:00:00.045 FRC: Some settings have been reset (4)
00:00:00.048 SPI: Hardware using GPIO12(CLK), GPIO11(MOSI) and GPIO13(MISO)
00:00:00.050 I2C: Bus1 using GPIO20(SCL) and GPIO19(SDA)
00:00:00.668 TFS: File 'mcp23x.dat' not found
00:00:00.679 Project tasmota - Tasmota7Zoll-8048S Version 13.3.0.1(lvgl-haspmota)-2_0_14(2023-12-21T19:49:30)
00:00:01.321 WIF: Connecting to AP1 .....
00:00:02.166 WIF: Connected
00:00:02.418 HTP: Web server active on Tasmota7Zoll-8048S with IP addres....
00:00:02.569 CMD: UrlFetch https://raw.githubusercontent.com/arendst/Tasmota/development/tasmota/berry/haspmota/haspmota_demo.tapp
00:00:02.583 RSL: RESULT = {"Command":"Unknown"}
02:15:27.394 RSL: INFO1 = {"Info1":{"Module":"SUNTON7","Version":"13.3.0.1(lvgl-haspmota)","FallbackTopic":"cmnd/DVES_8F981C_fb/","GroupTopic":"cmnd/tasmotas/"}}
02:15:27.406 RSL: INFO2 = {"Info2":{"WebServerMode":"Admin","Hostname":"Tasmota7Zoll-8048S","IPAddress"......st1"}}
02:15:27.418 RSL: INFO3 = {"Info3":{"RestartReason":{"Exception":28,"Reason":"LoadProhibited","EPC":"420863e2","EXCVADDR":"0000000c","CallChain"["420863df","420869cb","42075131","4207e1fd","4209ed28","4207e21d","420b84c6","420c3944","420272b6"]},"BootCount":12}}
02:15:27.450 RSL: RESULT = {"POWER":"OFF"}
02:15:27.451 RSL: POWER = OFF
02:15:30.914 QPC: Reset
02:15:31.910 RSL: STATE = {"Time":"2024-01-22T02:15:31","Uptime":"0T00:00:08","UptimeSec":8,"Heap":229,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":20,"MqttCount":0,"POWER":"OFF","Wifi":{"AP":1,"SSId":"Phradio","BSSId":"2C:91:AB:1E:A2:C8","Channel":11,"Mode":"11n","RSSI":46,"Signal":-77,"LinkCount":1,"Downtime":"0T00:00:03"}}
02:15:34.356 CMD: UrlFetch https://raw.githubusercontent.com/arendst/Tasmota/development/tasmota/berry/haspmota/haspmota_demo.tapp
02:15:34.370 RSL: RESULT = {"Command":"Unknown"}
```

```
Program Version | 13.3.0.1(lvgl-haspmota)
ESP Chip Id	9410588 (ESP32-S3 v0.1)
Flash Chip Id	0x1840C8 (QIO)
Flash Size	16384 KB
Program Flash Size	16384 KB
Program Size	2245 KB
Free Program Space	634 KB
Free Memory	224.2 KB
```

**Comment by ChrisVonFrosta** [comment]
i have this datasheets:
[7.0_module_ssd1963 V4.6.pdf](https://github.com/arendst/Tasmota/files/14003651/7.0_module_ssd1963.V4.6.pdf)
[ESP32-8048S070_SSD1963_RGB.pdf](https://github.com/arendst/Tasmota/files/14003652/ESP32-8048S070_SSD1963_RGB.pdf)
[Touch IC XPT2046.pdf](https://github.com/arendst/Tasmota/files/14003656/Touch.IC.XPT2046.pdf)
[SSD1963.pdf](https://github.com/arendst/Tasmota/files/14003657/SSD1963.pdf)

example this lib-files:
[lib.zip](https://github.com/arendst/Tasmota/files/14003643/lib.zip)

**Comment by ChrisVonFrosta** [comment]
It seems that the diplay has more than 16 Bits for RGB. Can Tasmota handle this?

**Comment by ChrisVonFrosta** [comment]
it seems that i download/provide wrong dataheets. 
this should the product corresponding files (i had to cut off to big (demo) files becauce file upload limit):
[ESP32-8048S070 Specifications-EN.pdf](https://github.com/arendst/Tasmota/files/14004518/ESP32-8048S070.Specifications-EN.pdf)
[7.0inch_ESP32-8048S070.zip](https://github.com/arendst/Tasmota/files/14004242/7.0inch_ESP32-8048S070.zip)

original zip download [link ](http://pan.jczn1688.com/directlink/1/ESP32%20module/7.0inch_ESP32-8048S070.zip?spm=a2g0o.detail.1000023.1.5e3drX2DrX2DuJ&file=7.0inch_ESP32-8048S070.zip) (with demo and lib code ) from product page.

**Comment by ChrisVonFrosta** [comment]
i reinstalled latest `Tasmota32-S3 (QIO Flash/PSRAM)` over web installer

now after reboot my console log is:
```
00:00:00.743 BRY: Berry initialized, RAM used 5684 bytes
00:00:00.782 I2C: GT911 found at 0x5d
00:00:00.901 TI: GT911
00:00:01.350 DSP: Configured display 'ILI9488'
00:00:01.358 Project tasmota - Tasmota7Zoll-8048S Version 13.3.0.3(lvgl-haspmota)-2_0_14(2024-01-21T21:28:39)
00:00:01.361 LVG: LVGL initialized
00:00:02.024 TAP: Loaded Tasmota App 'haspmota_demo.tapp'
00:00:04.005 BRY: failed to run compiled code 'value_error' - value needs to be a number
```

my actually `display.ini`:
```
:H,EK9716,800,480,16,RGB,41,40,39,42,2,15,7,6,5,4,9,46,3,8,16,1,14,21,47,48,45,14
:S,2,1,1,0,40,20
:0,00
:1,01
:2,02
:3,03
:TI1,5d,*,*,-1,38
#

`Option A` is set to `3`
```

**Comment by sfromis** [comment]
When having a severe bootloop issue, Tasmota may disable features to try to get to booting without quicky crashing again.
> `00:00:00.045 FRC: Some settings have been reset (4)`

Such issues should be resolved before trying to push on with anything else, and may explain why `UrlFetch` could not be recognized.

And until booting with PSRAM, better forget about seeing it as viable for supporting a display. Maybe you need `qio_opi`

> `00:00:01.350 DSP: Configured display 'ILI9488'`

Not sure what you're mixing up. Maybe you have unsuitable gpio configs?

**Comment by ChrisVonFrosta** [comment]
i ended up with `Tasmota32-S3 (QIO Flash/OPI PSRAM` and this display.ini:
```
:H,EK9716BD4,800,480,16,RGB,41,40,39,42,2,15,7,6,5,4,9,46,3,8,16,1,14,21,47,48,45,14
:S,2,1,1,0,40,20
:0,00
:1,01
:2,02
:3,03
:TI1,5d,*,*,-1,38
#
```

But i have no stattingpoint for Initialiozation "I:". There are some exapmples but i can't find any documentation for parameters. 
the only thing i found:
![grafik](https://github.com/arendst/Tasmota/assets/16272493/dd0df85a-e7b1-4c6b-9779-5dafe8ec3ef7)

console output after restart:
```
00:00:00.001 CMD: Using USB CDC
00:00:00.001 HDW: ESP32-S3 v0.1 (PSRAM)
00:00:00.020 UFS: FlashFS mounted with 12460 kB free
00:00:00.031 CFG: Loaded from File, Count 41
00:00:00.037 QPC: Count 1
00:00:00.040 SPI: Hardware using GPIO12(CLK), GPIO11(MOSI) and GPIO13(MISO)
00:00:00.040 I2C: Bus1 using GPIO20(SCL) and GPIO19(SDA)
00:00:00.645 TFS: File 'mcp23x.dat' not found
00:00:00.675 BRY: Berry initialized, RAM used 5684 bytes
00:00:00.691 I2C: GT911 found at 0x5d
00:00:00.810 TI: GT911
00:00:01.300 DSP: Configured display 'EK9716BD4'
00:00:01.306 Project tasmota - Tasmota7Zoll-8048S Version 13.3.0.3(lvgl-haspmota)-2_0_14(2024-01-21T21:28:39)
00:00:01.310 LVG: LVGL initialized
00:00:01.954 TAP: Loaded Tasmota App 'haspmota_demo.tapp'
00:00:03.369 BRY: failed to run compiled code 'value_error' - value needs to be a number
00:00:06.000 WIF: Connecting to AP1 
```

**Reply by ChrisVonFrosta** [reply]
the headerline should be correct. i found this:
![grafik](https://github.com/arendst/Tasmota/assets/16272493/dc941922-0031-451c-aa1c-c52996054104)
from [here](https://github.com/moononournation/Arduino_GFX/wiki/Dev-Device-Declaration#esp32-8048s070)

**Reply by sfromis** [reply]
AFAICT, what can or should be done in `:I` cannot be described in Tasmota docs, as this is entirely up to your display controller chip, and how to initialize it. While existing setups in displaydesc or autoconf can be "inspirational", it also serves to show how different those initializations can be.

However, the setup for the exact same controller chip in the [original Sunton](https://github.com/tasmota/autoconf/blob/main/raw/esp32s3/SunTon-S3-Display-7inch/display.ini) does not include any usage of `:I`. Are you positive that yours is not just a clone of Sunton? By "clone", I do not mean 100% exact, but the usual "China-style" of "very, very similar, but a bit different".

Everything you have in what you "ended up with" is the same as in the "real" Sunton, except for you omitting the `:V` line, and having some different gpio numbers. How much works?

**Reply by ChrisVonFrosta** [reply]
yes, after sleeping a while i checked what `I:` do. Is the `display.ini` format an "old style" script? BTW: Is it fundamentally possible to initialize the display using Berry?


i have no idea how simliar this board is to original Sunton. compared with your [link ](https://github.com/tasmota/autoconf/blob/main/raw/esp32s3/SunTon-S3-Display-7inch/display.ini) (thanks for that) the pinmapping seems to be different a bit
`:H,EK9716BD4,800,480,16,RGB,41,40,39,42,2,15,7,6,5,4,9,46,3,8,16,1,14,21,47,48,45,14`

i tried this [init part](https://github.com/arendst/Tasmota/discussions/20527#discussioncomment-8197340) with this result:
https://github.com/arendst/Tasmota/assets/16272493/88a9dba0-95d7-4c52-a0c2-8926f7455312

now i tried the complete [Sunset init](https://github.com/tasmota/autoconf/blob/main/raw/esp32s3/SunTon-S3-Display-7inch/display.ini) without care about the differents in portmapping and have this better result:
https://github.com/arendst/Tasmota/assets/16272493/d9101693-7bc9-4ec7-bb99-07e0c42dde7d
![grafik](https://github.com/arendst/Tasmota/assets/16272493/d8191a06-3235-4a87-9904-479fceed8fa4)


Looks good but the display size does not fit. Or does HASPmota Demo not fitting to larger displays?
And still HASPmota error:
```
00:00:00.001 CMD: Using USB CDC
00:00:00.001 HDW: ESP32-S3 v0.1 (PSRAM)
00:00:00.063 UFS: FlashFS mounted with 12460 kB free
00:00:00.084 CFG: Loaded from File, Count 59
00:00:00.090 QPC: Count 1
00:00:00.093 SPI: Hardware using GPIO12(CLK), GPIO11(MOSI) and GPIO13(MISO)
00:00:00.093 I2C: Bus1 using GPIO20(SCL) and GPIO19(SDA)
00:00:00.699 TFS: File 'mcp23x.dat' not found
00:00:00.765 BRY: Berry initialized, RAM used 5684 bytes
00:00:00.793 I2C: GT911 found at 0x5d
00:00:00.912 TI: GT911
00:00:01.379 DSP: Configured display 'EK9716_sunton'
00:00:01.387 Project tasmota - Tasmota7Zoll-8048S Version 13.3.0.3(lvgl-haspmota)-2_0_14(2024-01-21T21:28:39)
00:00:01.391 LVG: LVGL initialized
00:00:02.116 TAP: Loaded Tasmota App 'haspmota_demo.tapp'
00:00:04.503 BRY: failed to run compiled code 'value_error' - value needs to be a number
00:00:07.926 WIF: Connecting to AP1 
```

**Reply by sfromis** [reply]
That `:I` (or `display.ini` at all) is not part of the scripter, it is a separate display descriptor specification. It works fine with Berry (LVGL/HASPmota), or plain commands without having the scripter feature in the build.

And yes, when you correct for the difference in pin numbers, it appears to work just like with my Sunton ESP32-8048S070. Same colors and scaling. That demo is made for a much smaller display, hence it sits in the corner when getting 800x480 instead of 240x320. Thus, I'd say that the display size works fine, but having "nice" pages means adapting to the intended screen resolution and dimensions.

The error message is not really important, looks like only some small part of the demo may be inop
> `00:00:04.503 BRY: failed to run compiled code 'value_error' - value needs to be a number`

Still "deserves" debugging and fixing, of course. I'm also getting it.

In short, you really have a pretty close Sunton clone, just with different pins.

Here's a quick-and-dirty `colorgradients.jsonl` test setup to verify that color pins are in the right order, otherwise the gradients would have bad stripes.
```
{"page":0}
{"id":101,"obj":"btn","x":  8,"y":8,"w":48,"h":24,"action":"prev","bg_color":"#1fa3ec","radius":10,"border_side":1,"text":"\uF053","text_font":"montserrat-20"}
{"id":102,"obj":"btn","x": 88,"y":8,"w":48,"h":24,"action":"back","bg_color":"#1fa3ec","radius":10,"border_side":1,"text":"\uF015","text_font":"montserrat-20"}
{"id":103,"obj":"btn","x":168,"y":8,"w":48,"h":24,"action":"next","bg_color":"#1fa3ec","radius":10,"border_side":1,"text":"\uF054","text_font":"montserrat-20"}
{"page":1}
{"id":0,"bg_color":"#000000","bg_grad_color":"#ff0000","bg_grad_dir":1,"text_color":"#FFFFFF"}
{"id":1,"obj":"label","x":32,"y":48,"w":140,"text":"Red","align":0,"text_font":"montserrat-20"}
{"page":2}
{"id":0,"bg_color":"#000000","bg_grad_color":"#00ff00","bg_grad_dir":1,"text_color":"#FFFFFF"}
{"id":1,"obj":"label","x":32,"y":48,"w":140,"text":"Green","align":0,"text_font":"montserrat-20"}
{"page":3}
{"id":0,"bg_color":"#000000","bg_grad_color":"#0000ff","bg_grad_dir":1,"text_color":"#FFFFFF"}
{"id":1,"obj":"label","x":32,"y":48,"w":140,"text":"Blue","align":0,"text_font":"montserrat-20"}
```

**Reply by sfromis** [reply]
When it comes to display initialization via Berry, I've not tried this, but the `display` module does include a `start` method to initialize uDisplay including a string with content similar to `display.ini`
https://tasmota.github.io/docs/Berry/#display-module

**Comment by ChrisVonFrosta** [comment]
the screen build/reaction time is not great, i can see a part from next view "sliding in" from side.
Is the reaction slower when using HASPmota instead of building pages with berry script?

**Reply by sfromis** [reply]
HASPmota is already doing it with Berry, but I think that the slow transitions (which I also see) happens deeper in the hierarchy, and the coding was originally done for much smaller displays. And HASPmota is not doing the pixel-level handling, that is delegated to LVGL and uDisplay.

**Reply by Jason2866** [reply]
It is a slow display.

**Reply by sfromis** [reply]
It may also be a question of how much the code has been tuned for the RGB display interface support of the ESP32-S3, like using frame buffers in an efficient way.

**Reply by ChrisVonFrosta** [reply]
> It is a slow display.

hm. that can't be the reason. Those display boards (i got 4 different) came with an demo application and it reacted remarkable faster.

**Comment by ScanLabs** [comment]
> ChrisVonFrosta

Hello ChrisVonFrosta, I'm on the same situation as you were ... have you a setup to share to let Tasmota LVGL working with ESP32-8048S070 ? I tried multiple sources but so far 0 results.

Thanks,
Paolo

**Reply by sfromis** [reply]
For me, it "works" by selecting the auto-configuration "SunTon S3 Display 7inch", after having installed or upgraded to a binary supporting LVGL+HASPmota and PSRAM, `tasmota32s3-qio_opi.bin` from https://github.com/tasmota/install/blob/firmware/firmware/unofficial/ (also available in the Tasmota Web Installer).

However "works" in quotes as the display does not get stable. You can improve on that by doing a build of `tasmota32s3-qio_opi_120` (no precompiled build available).

**Reply by ScanLabs** [reply]
Thanks  sfromis, I follow your suggestions and indeed now I can see the tasmota logo and may pages.json is shown.
I have a problem maybe you already managed.
Screen is mirrored on x. I tried with :R option without any success.
Any clue on that ?
tnx
paolo

**Reply by sfromis** [reply]
With a HASPmota test case, I'm seeing text rendered normally, no mirroring.

I've not tried adjusting mirroring, but AFAIK, you can use the 4 lines `:0` .. `:3` to adjust flags for flip/swap per rotation setting. The existing set of entries are for a particular set of orientations.
See: https://tasmota.github.io/docs/Universal-Display-Driver/#descriptor-file
